File: cabal/tests/HelloWorldTests.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Data.Time.Clock (getCurrentTime)
import Data.Time.Format (defaultTimeLocale, formatTime)
import Data.Time.LocalTime (utcToLocalZonedTime)
import Hello.World (printDate, printGreeting)
import Test.Tasty (TestTree, defaultMain, testGroup)
import Test.Tasty.HUnit (assertBool, testCase)

formatErrorMessage :: String -> String -> IO String
formatErrorMessage expected actual = do
    now <- getCurrentTime >>= utcToLocalZonedTime
    return $
        "\nTest run at: "
            ++ show now
            ++ "\nExpected : "
            ++ expected
            ++ "\nActual   : "
            ++ actual
            ++ "\n"

printGreetingWithoutNameTest :: TestTree
printGreetingWithoutNameTest = testCase "Print Greeting without Name Test" $ do
    let args = []
    let expectedOutput = "Hello World from Cabal !"
    let actualOutput = printGreeting args
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

printGreetingWithNameTest :: TestTree
printGreetingWithNameTest = testCase "Print Greeting with Name Test" $ do
    let args = ["Alter", "Sabeh"]
    let expectedOutput = "Hello Alter Sabeh from Cabal !"
    let actualOutput = printGreeting args
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

printDateTest :: TestTree
printDateTest = testCase "Print Date Test" $ do
    now <- getCurrentTime >>= utcToLocalZonedTime
    let formattedDate = formatTime defaultTimeLocale "%B %d, %Y at %I:%M %p" now
    let expectedOutput = "Today is " ++ formattedDate
    actualOutput <- printDate
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

tests :: TestTree
tests =
    testGroup
        "Hello.World Tests"
        [printGreetingWithoutNameTest, printGreetingWithNameTest, printDateTest]

main :: IO ()
main = defaultMain tests

File: gradle/packages/hello-world/src/test/kotlin/com/app/hello/world/HelloWorldTests.kt
package com.app.hello.world

import org.apache.commons.lang3.time.DateFormatUtils
import kotlin.test.Test
import kotlin.test.assertEquals

class HelloWorldTests {
    private fun formatErrorMessage(
        expected: String,
        actual: String,
    ): String {
        val now = System.currentTimeMillis()
        val formattedDate = DateFormatUtils.format(now, "yyyy-MM-dd'T'HH:mm:ss")
        return """
            |${"\u001B[31m"}
            |Test run at: $formattedDate
            |Expected : $expected
            |Actual : $actual
            |
            |${"\u001B[00m"}
            """.trimMargin()
    }

    @Test
    fun `print greeting without name test`() {
        val args = emptyArray<String>()
        val expectedOutput = "Hello World from Gradle !"
        val actualOutput = printGreeting(args)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    @Test
    fun `print greeting with name test`() {
        val args = arrayOf("Alter", "Sabeh")
        val expectedOutput = "Hello Alter Sabeh from Gradle !"
        val actualOutput = printGreeting(args)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    @Test
    fun `print current date test`() {
        val now = System.currentTimeMillis()
        val formattedDate = DateFormatUtils.format(now, "MMMM dd, yyyy 'at' hh:mm a")
        val expectedOutput = "Today is $formattedDate"
        val actualOutput = printDate()
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }
}
